### ISSUE TITLE
[proposal+

### ISSUE THREAD

**Issue by v01d** [koreader/koreader#4553]
* KOReader version:
* Device:

##### `crash.log` (if applicable)
`crash.log` is a file that is automatically created when KOReader crashes. It can
normally be found in the KOReader directory:

* `koreader/` directory for Kindle
* `.adds/koreader/` directory for Kobo
* `applications/koreader/` directory for Pocketbook

Please try to include the relevant sections in your issue description.
You can upload the whole `crash.log` file on GitHub by dragging and
dropping it onto this textbox, or you can use one of these paste services.
Please don't forget to set a high expiration date if you do.
https://pastebin.com https://slexy.org https://paste2.org https://fpaste.org
http://ix.io https://paste.kde.org https://paste.debian.net.

**Comment by Frenzie** [comment]
Sounds like a cool idea. Personally I'm not too interested in finger writing though. Btw, on what kind of device is it? There's a pretty cool drawing app by @baskerville for Kobo already.

**Comment by v01d** [comment]
I mean with a stylus/pen. I have a Sony PRS-T2 reader with IR display. You can use any pointy thing to write. I don't expect much precision, but for capacitive displays, maybe an appropriate pen could be good enough for not-so-small writing.

**Comment by v01d** [comment]
PS: I'm thinking something along the lines of what the remarkable ereader does.

**Comment by v01d** [comment]
From the architectural point of view, should this be added as an "app" or as a plugin? I see the text editor is actually a plugin, but that feels weird. I imagine a plugin solves the problem of how to access the functionality whereas an app would require some form of UI/input extension, right?

For capturing and displaying traces, should I use the SVG library?

**Comment by poire-z** [comment]
> I see the text editor is actually a plugin, but that feels weird

The text editor plugin is just a simple UI "app" that brings some menu items and deal with selecting/opening/saving files. It uses the [InputDialog](https://github.com/koreader/koreader/blob/master/frontend/ui/widget/inputdialog.lua) widget that provides and does most of the text editing work (with the help of a few other widgets line InputText). That InputDialog is also used by other plugins or core features (text input in dialogs, wikipedia search term, editing highlights...)
You can read more about that in a discussion we had (well, at least _my_ point of view :) in https://github.com/koreader/koreader/issues/4073#issuecomment-404656687 and followup posts.

The things in frontend/apps/ are the most essential apps, that makes koreader a reader and a filemanager :) (opdscatalog and cloudstorage could very well be plugins.)
Also, you can disable plugins that does not interest you (I do disable most), and plugins are loaded in some kind of sandbox, so they won't crash koreader in case they do bad things.

So, you could first go with a plugin for safety and not invading the core, as I think a sketch app that generates images would not provide any services to the core.
But if you feel like finger writting could be a proper input method (for wikipedia seach, renaming files) - which would mean you can translate what's written into unicode chars :), you could make a frontend/ui/widget/handwritintinputmethod.lua that would provide this service. But you would still have a Sketch plugin that would use this widget, but it would provide the UI/brush size color selection/save to PNG/SVG things.

Note that there is no support (yet?) for SVG in KOReader. Except a small nanosvg library used to do limited decoding of SVG images found in EPUB Files.

Also, I'm not sure our low level lua input module that translate touch hardware event to "swipe north", "touch x/y", "pan 40px south" would not be a blocker for what you want to do.
And if you need to bypass it, well, it might be necessary to kill/stop/bypass that KOreader feature, and may be to exit koreader so you plug to the linux events, so... launching an alternative app :| which would not make it part of KOReader :| It would be better to clarify that first before you jump into the UI work...

**Comment by Frenzie** [comment]
Like @poire-z said, architecturally speaking most things should be a plugin when reasonably possible. In practice something like the Reader could be conceived as a plugin I suppose, but besides predating plugins in KOReader I don't think that would make much sense.

> And if you need to bypass it, well, it might be necessary to kill/stop/bypass that KOreader feature, and may be to exit koreader so you plug to the linux events

This is a case where changing the core to be more suitable for the plugin would probably benefit the program as a whole. A gesture manager is a long-standing request/target, and perhaps it could also suppress GestureDetector on demand. (NB Just thinking out loud.)

**Comment by v01d** [comment]
Thanks for the valuable input. I'm concerned about the lack of ability of listening to raw input events if koreader is reading this. Is this a problem of exclusive access to the device? Can't I simply open it for reading myself as well?

Indeed I could make it completely independent of koreader, but I feel that many things are already solved with koreader so I would be reinventing many little wheels.

**Comment by Frenzie** [comment]
It's probably easiest to read through the code a little, otoh these two files are probably the most relevant.

* [frontend/device/input.lua](https://github.com/koreader/koreader/blob/812e595608fcb4d7d6e7ac87fb221942fc15af8b/frontend/device/input.lua)
* [frontend/device/gesturedetector.lua](https://github.com/koreader/koreader/blob/812e595608fcb4d7d6e7ac87fb221942fc15af8b/frontend/device/gesturedetector.lua)

https://github.com/koreader/koreader/blob/812e595608fcb4d7d6e7ac87fb221942fc15af8b/frontend/device/input.lua#L427-L453

So we do have a raw list of events but it's filtered through gesture detector.

https://github.com/koreader/koreader/blob/812e595608fcb4d7d6e7ac87fb221942fc15af8b/frontend/device/input.lua#L508

It might be as simple as just storing a copy of gesture detector and making it a dummy while the drawing plugin is displayed, especially as a relatively quick initial proof of concept. But however it's implemented, you basically want to respond to all touch events without the normal distinction between tap, swipe, hold, etc. that pervades the program.

**Comment by Frenzie** [comment]
Yeah, mouse up/down actually generates fake touch events, see [here](https://github.com/koreader/koreader-base/blob/afd0c6c5a6d50e15ecfede981f5c8bcb768ad43a/ffi/SDL2_0.lua#L274-L289).

**Comment by Frenzie** [comment]
Oh btw, there's the [wbuilder](https://github.com/koreader/koreader/blob/812e595608fcb4d7d6e7ac87fb221942fc15af8b/tools/wbuilder.lua) to iterate on UX stuff quickly. It depends a bit on what you want to start with.

**Comment by pazos** [comment]
I think that should be possible to ignore events in KOReader while we are in "raw_input_mode", a la [screen_saver_mode](https://github.com/koreader/koreader/blob/master/frontend/ui/uimanager.lua#L193-L195) and process raw linux events (real events in kobo, kindle, sony, cervantes and fake ones in android and pocketbook)

**Comment by Frenzie** [comment]
PS I meant working fully within KOReader's event abstraction, but without the other touch gestures for this specific purpose.

**Comment by mapinguari** [comment]
Hi all, was there any progress made on this? I was also thinking about trying to implement this and wonder if there is any work that has already been completed

**Comment by Frenzie** [comment]
Not that I know of. ^_^

**Comment by fgheng** [comment]
Hello，does it support handwrite now?

**Comment by Eeems** [comment]
> Hello，does it support handwrite now?

Since this is still open, no.

**Comment by Iey4iej3** [comment]
There is a recent cross-platform app https://github.com/adil192/saber based on flutter, and I wonder how difficult it is to adapt these codes?

**Comment by Frenzie** [comment]
In the context of something so different I doubt adapt could (or even should) mean much more than vaguely inspired by.

**Comment by NiLuJe** [comment]
You won't find *anything* to *adapt* anyway, anything related to this (or... anything else, really), will have to be brand new and specific for KOReader.

**Comment by Frenzie** [comment]
Eh, if it were Plato for example you might be able to "adapt" (read: use as documentation) how it reads the input or something. ;-)

**Comment by NiLuJe** [comment]
I'm pretty sure that's not what was implied there, though ;o). But, yes, documentation is the extent of what might be feasible ;).

**Comment by Frenzie** [comment]
Sure, but you said "anything else" xD

**Comment by NiLuJe** [comment]
Oh, I meant that in a sense of any other features, not other... "sources of inspiration" ;o).

**Comment by Glavnokoman** [comment]
came here looking for adding "freehand notes" to the documents, not exactly what author meant. +1 For the feature request though.

**Comment by slip-per** [comment]
Yes! I had this idea also. I just replaced the battery on an old 3rd gen kindle I got online. I like the manual keyboard, not as laggy as the touch screen keyboard. I'm looking for a 'notes' type plugin for koreader too so I can use the jailbroken gen 3 as more of a journal. Haven't found anything so far and don't write code myself... so will be lurking the forums until someone makes one

**Comment by MrDaisyBates** [comment]
👀
